math/pythagoras.cpp

```diff
@@ -0,0 +1,94 @@
+
+/**
+ * Copyright 2020 @author beqakd
+ * @file
+ *
+ * A basic implementation of 3d distance between two point.
+ */
+
+#include <cassert>   // For testing(assert)
+#include <cmath>     // For std::pow(), std::sqrt() and so on
+#include <iostream>  // For io operations
+
+/**
+ * The distance between two points in a three dimensional - 3D - coordinate
+ * system. For this we have 3 point x, y, z.
+ * Also we created method int distance which takes points and returns distance.
+ */
+template <typename T> class point {
+ private:
+    T x;
+    T y;
+    T z;
+
+ public:
+    /**
+    * We have two constructor. First is for 3D point. it takes T type 3 value
+    * and save it to our calss. Second one takes only 2 T type value and third
+    * one is defoult 0 and save it into the class too.
+    */
+    point(T a, T b, T c) : x(a), y(b), z(c) {}  // Consructor for 3D points
+    point(T a, T b) : x(a), y(b), z(0) {}  // Second constructor for 2D points
```

Comment:
i think we do not need this explicitly, because a point in 2D will have Z=0 (always) so we can default z=0 in 3D point constructor by doing this the third argument that is Z becomes optional if the uses passes a 3D point then it will replace Z=0 with the value that is given by user and it will work for the 3D point in this way and if the user does not passes the third argument then it automatically becomes a point in 2D (z=0) and the program will behave accordingly